src/hip_memory.cpp

```diff
@@ -1573,15 +1573,60 @@ int getByteSizeFromFormat(const hipChannelFormatDesc& desc){
     return byteSize;
 }
 
-hipError_t ihipMemcpy3D(const struct hipMemcpy3DParms* p, hipStream_t stream, bool isAsync) {
-    hipError_t e = hipSuccess;
-    if(p) {
-        size_t dstByteSize, srcByteSize, copyWidth, copyHeight, copyDepth, widthInBytes, srcPitch, dstPitch, srcYsize, dstYsize;
-        size_t srcXoffset, srcYoffset, srcZoffset, dstXoffset, dstYoffset, dstZoffset;
-        size_t srcWidth, srcHeight, srcDepth, dstWidth, dstHeight, dstDepth;
+int getByteSizeFromArrayFormat(const hipArray_Format format){
+   int byteSize =0;
+   switch (format) {
+       case HIP_AD_FORMAT_UNSIGNED_INT8:
+          byteSize = sizeof(uint8_t);
+          break;
+       case HIP_AD_FORMAT_UNSIGNED_INT16:
+          byteSize = sizeof(uint16_t);
+          break;
+       case HIP_AD_FORMAT_UNSIGNED_INT32:
+          byteSize = sizeof(uint32_t);
+          break;
+       case HIP_AD_FORMAT_SIGNED_INT8:
+          byteSize = sizeof(int8_t);
+          break;
+       case HIP_AD_FORMAT_SIGNED_INT16:
+          byteSize = sizeof(int16_t);
+          break;
+       case HIP_AD_FORMAT_SIGNED_INT32:
+          byteSize = sizeof(int32_t);
+          break;
+       case HIP_AD_FORMAT_HALF:
+          byteSize = sizeof(_Float16);
+          break;
+       case HIP_AD_FORMAT_FLOAT:
+          byteSize = sizeof(float);
+          break;
+       default:
+          break;
+   }
+   return byteSize;
+}
 
-        void* srcPtr, *dstPtr;
-        bool copyWidthUpdate= false;
+hipError_t ihipMemcpy3D(const struct hipMemcpy3DParms* p, const HIP_MEMCPY3D* pCopy, hipStream_t stream, bool isAsync, bool isDrv) {
+    hipError_t e = hipSuccess;
+    size_t dstByteSize, srcByteSize, copyWidth, copyHeight, copyDepth, widthInBytes, srcPitch, dstPitch, srcYsize, dstYsize;
+    size_t srcXoffset, srcYoffset, srcZoffset, dstXoffset, dstYoffset, dstZoffset;
+    size_t srcWidth, srcHeight, srcDepth, dstWidth, dstHeight, dstDepth;
+
+    void* srcPtr, *dstPtr;
+    bool copyWidthUpdate= false;
+    if(isDrv) {
+        if(pCopy == nullptr) return hipErrorInvalidValue;
+        copyDepth = pCopy->Depth;
+        copyHeight = pCopy->Height;
+        copyWidth = pCopy->WidthInBytes;
+        dstXoffset = pCopy->dstXInBytes;
+        dstYoffset = pCopy->dstY;
+        dstZoffset = pCopy->dstZ;
+        srcXoffset = pCopy->srcXInBytes;
+        srcYoffset = pCopy->srcY;
+        srcZoffset = pCopy->srcZ;
```

Comment:
Here and below same as mentioned in already merged #1798:
 I'd think about `hipMemcpy3DParms` and `HIP_MEMCPY3D` struct var on the stack for copying routines to avoid massively error-prone copying of fields.